In [1]:
################################
#
# Dataset: https://www.kaggle.com/tevecsystems/retail-sales-forecasting
#
# The data contains the sales and price of a product at the end of each day
# We will forecast the sales for next day, given the sales for the last few days
# Using LSTMs cells and RNN
#
# Author: Anurag Dwarakanath
################################

In [2]:
import tensorflow as tf
import pandas
import numpy as np

In [24]:
# Constants
DATAFILE_TRAIN = "/home/ubuntu/anurag/rnn/mock_kaggle_edit_train.csv" 

DATAFILE_VALIDATE = "/home/ubuntu/anurag/rnn/mock_kaggle_edit_validate.csv"

# The original data is here: https://www.kaggle.com/tevecsystems/retail-sales-forecasting
# The original data was simplified to consider only 'date', 'sales', 'price'
# However, we will consider only 'sales' & 'date'

TRAINED_MODEL_PATH = '/home/ubuntu/anurag/rnn/savedModel'

TIME_STEPS = 10 # i.e. look at the past 10 days and forecast
NUMBER_OF_DAYS_TO_FORECAST = 1 # for now we will only forecast the next day's sales

BATCH_SIZE=100

NUM_EPOCHS=100

LSTM_UNITS = 250

TENSORBOARD_LOGDIR = '/home/ubuntu/anurag/rnn/tensorboard_log'

In [25]:
#ingest the data
data_train = pandas.read_csv(DATAFILE_TRAIN) # data will be an array of dict. E.g. data['sales'][0]
data_validate = pandas.read_csv(DATAFILE_VALIDATE)

In [26]:
# let's print the first 5 rows of the data
data_train.head()

date  sales  price
0  1/1/2014      0   1.29
1  1/2/2014     70   1.29
2  1/3/2014     59   1.29
3  1/4/2014     93   1.29
4  1/5/2014     96   1.29

In [27]:
# Create the training & validation data

numTrainingData = len(data_train)
numValidationData = len(data_validate)

trainingData_date = data_train['date'][0:numTrainingData]
trainingData_sales = data_train['sales'][0:numTrainingData]
trainindData_price = data_train['price'][0:numTrainingData]

validationData_date = data_validate['date'][0: numValidationData]
validationData_sales = data_validate['sales'][0: numValidationData]
validationData_price = data_validate['price'][0: numValidationData]

In [45]:
trainingData_sales.head()

0     0
1    70
2    59
3    93
4    96
Name: sales, dtype: int64

In [29]:
print(len(trainingData_sales))
print(len(validationData_sales))

750
187


In [30]:
# Normalise the training and validation data.
# We use the (x - min)/(range) to normalise, where range = (max - min)
# This will put the values between 0 & 1
trainingData_sales_min = min(trainingData_sales)
trainingData_sales_max = max(trainingData_sales)
trainingData_sales_range = trainingData_sales_max - trainingData_sales_min
trainingData_sales_normalised = [(i - trainingData_sales_min)/trainingData_sales_range for i in trainingData_sales]

validationData_sales_normalised = [(i - trainingData_sales_min)/trainingData_sales_range for i in validationData_sales]
# Note the validation data uses the max & min values of Training Data

In [31]:
print('Min:', trainingData_sales_min)
print('Range:', trainingData_sales_max - trainingData_sales_min)

Min: 0
Range: 542


In [32]:
# we will create the set of sequences
trainingDataSequence_sales = np.zeros(shape=(len(trainingData_sales)-TIME_STEPS-NUMBER_OF_DAYS_TO_FORECAST+1, TIME_STEPS, 1))
targetDataSequence_sales = np.zeros(shape=(len(trainingData_sales)-TIME_STEPS-NUMBER_OF_DAYS_TO_FORECAST+1, NUMBER_OF_DAYS_TO_FORECAST))
start = 0
for i in range(TIME_STEPS, len(trainingData_sales)-NUMBER_OF_DAYS_TO_FORECAST+1):
    trainingDataSequence_sales[start,:,0]=trainingData_sales_normalised[start:i]
    targetDataSequence_sales[start] = trainingData_sales_normalised[i:i+NUMBER_OF_DAYS_TO_FORECAST]
    start=start+1

## MT: make sure the sequences do not miss out any data
# maybe create predictable data (1, 2, 3, 4) then put completely rubbish data (23434) at the end.. this should be picked up in training


## The format of the data sequences is as follows:

trainingDataSequence_sales = \[sales_day1 sales_day2 sales_day3 ... sales_day10\] 

targetDataSequence_sales = \[sales_day11\]

Although the trainingDataSequence_sales can be captured through a 2D array, we use 3D (i.e. shape = (num of sequence, time steps, 1) ) because the tensor input for the LSTM cell expects a 3D input. The third dimension is the *num of features*  of each data point (in our case, it is a single dimension).

In [33]:
# Print the first few sequences of the sales that have been created
[trainingDataSequence_sales[i, :, 0] for i in range(3)]

[array([0.        , 0.12915129, 0.10885609, 0.17158672, 0.17712177,
        0.26752768, 0.3302583 , 0.59225092, 0.23062731, 0.16236162]),
 array([0.12915129, 0.10885609, 0.17158672, 0.17712177, 0.26752768,
        0.3302583 , 0.59225092, 0.23062731, 0.16236162, 0.34686347]),
 array([0.10885609, 0.17158672, 0.17712177, 0.26752768, 0.3302583 ,
        0.59225092, 0.23062731, 0.16236162, 0.34686347, 0.22324723])]

In [34]:
#Print the first few target sales
[targetDataSequence_sales[i] for i in range(3)]

[array([0.34686347]), array([0.22324723]), array([0.24723247])]

###### Observe that the first sales target (0.34686347) becomes the last day sales of the next triainingDataSequence

i.e. targetDataSequence_sales\[i\] = targetDataSequence_sales\[i+1,TIME_STEPS-1,0\]

In [35]:
# since we are making each sequence independent to each other,
# the order of the sequences can be shuffled.
# Note that we need to ensure the target and the train sequences are in the same order
a = np.arange(len(targetDataSequence_sales))
np.random.shuffle(a)
trainingDataSequence_sales_shuffle = np.zeros(shape=(len(trainingData_sales)-TIME_STEPS-NUMBER_OF_DAYS_TO_FORECAST+1, TIME_STEPS, 1))
targetDataSequence_sales_shuffle = np.zeros(shape=(len(trainingData_sales)-TIME_STEPS-NUMBER_OF_DAYS_TO_FORECAST+1, NUMBER_OF_DAYS_TO_FORECAST))

loc=0
for i in a:
    trainingDataSequence_sales_shuffle[loc] = trainingDataSequence_sales[i]
    targetDataSequence_sales_shuffle[loc] = targetDataSequence_sales[i]
    loc+=1

trainingDataSequence_sales = trainingDataSequence_sales_shuffle
targetDataSequence_sales = targetDataSequence_sales_shuffle

In [36]:
# Build the seqeunces for validation
validationDataSequence_sales = np.zeros(shape=(len(validationData_sales) - TIME_STEPS - NUMBER_OF_DAYS_TO_FORECAST + 1, TIME_STEPS, 1 ))
validationDataSequence_sales_target = np.zeros(shape=(len(validationData_sales) - TIME_STEPS - NUMBER_OF_DAYS_TO_FORECAST + 1, NUMBER_OF_DAYS_TO_FORECAST))

start = 0
for i in range(TIME_STEPS, len(validationData_sales)-NUMBER_OF_DAYS_TO_FORECAST + 1):
    validationDataSequence_sales[start, :, 0] = validationData_sales_normalised[start:i]
    validationDataSequence_sales_target[start] = validationData_sales_normalised[i:i+NUMBER_OF_DAYS_TO_FORECAST]
    start+=1

In [37]:
#build the RNN model
tf.reset_default_graph()

inputSequencePlaceholder = tf.placeholder(dtype=tf.float32, shape=(None, TIME_STEPS, 1), name='inputSequencePlaceholder')
targetPlaceholder = tf.placeholder(dtype=tf.float32, shape=(None, NUMBER_OF_DAYS_TO_FORECAST), name='targetPlaceholder')

#cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_layers=1, num_units=5, name='LSTM_cell')
cell = tf.nn.rnn_cell.LSTMCell(num_units=LSTM_UNITS, name='LSTM_cell')


output, state = tf.nn.dynamic_rnn(cell=cell, inputs=inputSequencePlaceholder, dtype=tf.float32)
#output is of shape (Batch, time, num_units)

lastCellOutput = output[:, -1,:]

In [38]:
print('output:',output)
print('state:', state)
print('lastCellOutput:', lastCellOutput)

output: Tensor("rnn/transpose_1:0", shape=(?, 10, 250), dtype=float32)
state: LSTMStateTuple(c=<tf.Tensor 'rnn/while/Exit_3:0' shape=(?, 250) dtype=float32>, h=<tf.Tensor 'rnn/while/Exit_4:0' shape=(?, 250) dtype=float32>)
lastCellOutput: Tensor("strided_slice:0", shape=(?, 250), dtype=float32)


## Understanding the type and shape of **output** & **state**
the RNN gives two outputs for every sequence of input: a) output & b) state.

**output** is the per time step output (i.e. output from each time step and which is represented as a_t in our notes). Thus it has a shape of (Batch, 10, 25). We are interested only in the final time step output (i.e. Batch, 9, 250) and this is captured by lastCellOutput

the **state** is the value of the internal parameters of the LSTM that can be propogated to the next sequence if need be. The state is of type LSTMStateTuple which contains two tensors of shape (batch, 250) each.

In [39]:
#build the output layer from the RNN outputs
weights = tf.Variable(initial_value=tf.truncated_normal(shape=(LSTM_UNITS, NUMBER_OF_DAYS_TO_FORECAST)))
bias = tf.Variable(initial_value=tf.ones(shape=(NUMBER_OF_DAYS_TO_FORECAST)))

forecast = tf.add(x=tf.matmul(a=lastCellOutput, b=weights),y= bias, name='forecast_normalised_scale')

#tf.summary.scalar(tensor=forecast, name='forecast_normalised_scale')

# bring back the forecast to the original scale
forecast_originalScale = tf.add(x = forecast * trainingData_sales_range ,y = trainingData_sales_min, name='forecast_original_scale')

In [40]:
print(forecast)
print(forecast_originalScale)

Tensor("forecast_normalised_scale:0", shape=(?, 1), dtype=float32)
Tensor("forecast_original_scale:0", shape=(?, 1), dtype=float32)


the **forecast** is the layer that computes the final forecast. this is the y_hat in our notes. **forecast** will have a shape of (Batch, 1)

In [41]:
loss = tf.reduce_mean(tf.squared_difference(x=forecast, y=targetPlaceholder), name='loss_comp')

tf.summary.scalar(tensor=loss, name='loss')

<tf.Tensor 'loss:0' shape=() dtype=string>

In [42]:
optimizer = tf.train.AdamOptimizer(learning_rate=0.1)
minimize_step = optimizer.minimize(loss)

In [43]:
# print([n.name for n in tf.get_default_graph().as_graph_def().node])

In [44]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    sess.run(tf.local_variables_initializer())
    
    #Get the tensorboard writer
    tensorboard_writer = tf.summary.FileWriter(TENSORBOARD_LOGDIR, sess.graph)
    
    #merge all the summary ops
    all_summary_ops = tf.summary.merge_all()
    
    #start the training
    numSteps=0
    for e in range(NUM_EPOCHS):
        print('starting training for epoch:', e+1)
        
        startLocation=0
        iteration=0
        for iteration in range(int(len(targetDataSequence_sales)/BATCH_SIZE)):
            print('epoch:', e+1, ' iteration:', iteration+1)
            trainingBatchInput = trainingDataSequence_sales[startLocation:startLocation+BATCH_SIZE, :, :]
            trainingBatchTarget = targetDataSequence_sales[startLocation:startLocation+BATCH_SIZE]
            
            _, lsBatch, forecastBatch, forecastBatch_originalScale, summary_values = sess.run([minimize_step, loss, forecast, forecast_originalScale, all_summary_ops], feed_dict={inputSequencePlaceholder: trainingBatchInput,
                                                                                            targetPlaceholder: trainingBatchTarget})
            
            tensorboard_writer.add_summary(summary_values, numSteps)
            numSteps+=1
            
            if ((iteration+1) % 1 == 0):
                print('got a loss of:', lsBatch)
                print('the forecast of first 5 normalised are:', forecastBatch[0:5])
                print('while the actuals were normalised     :', trainingBatchTarget[0:5])
                print('the forecast of first 5 orignal scale are:', forecastBatch_originalScale[0:5])
                print('while the actuals were original scale     :', trainingBatchTarget[0:5]* trainingData_sales_range + trainingData_sales_min)
            
            startLocation+=BATCH_SIZE
        
        #pick up the training data that does not fit into BATCH_SIZE
        if (len(targetDataSequence_sales) > startLocation):
            print('epoch:', e+1, ' iteration:', iteration+1)
            trainingBatchInput = trainingDataSequence_sales[startLocation:len(targetDataSequence_sales), :, :]
            trainingBatchTarget = targetDataSequence_sales[startLocation:len(targetDataSequence_sales)]
            
            _, lsBatch, forecastBatch, forecastBatch_originalScale = sess.run([minimize_step, loss, forecast, forecast_originalScale], feed_dict={inputSequencePlaceholder: trainingBatchInput,
                                                                                            targetPlaceholder: trainingBatchTarget})
            
            print('got a loss of:', lsBatch)
            print('the forecast of first 5 normalised are:', forecastBatch[0:5])
            print('while the actuals were normalised     :', trainingBatchTarget[0:5])
            print('the forecast of first 5 orignal scale are:', forecastBatch_originalScale[0:5])
            print('while the actuals were original scale     :', trainingBatchTarget[0:5]* trainingData_sales_range + trainingData_sales_min)
    
    
        #end of 1 epoch. Perform validation
        totalValidationLoss = 0
        startLocation=0
        print('starting validation')
        for iter in range(len(validationDataSequence_sales)//BATCH_SIZE):
            validationBatchInput = validationDataSequence_sales[startLocation: startLocation+BATCH_SIZE, :, :]
            validationBatchTarget = validationDataSequence_sales_target[startLocation: startLocation+BATCH_SIZE]
            
            validationLsBatch, validationForecastBatch, validationForecastBatch_originalScale = sess.run([loss, forecast, forecast_originalScale], feed_dict={inputSequencePlaceholder: validationBatchInput,
                                                                                              targetPlaceholder: validationBatchTarget})
            
           
            startLocation+=BATCH_SIZE
            totalValidationLoss+= validationLsBatch
            
            print('first five predictions:', validationForecastBatch[0:5])
            print('first five actuals    :', validationBatchTarget[0:5])
            print('the forecast of first 5 orignal scale are:', validationForecastBatch_originalScale[0:5])
            print('while the actuals were original scale     :', validationBatchTarget[0:5]* trainingData_sales_range + trainingData_sales_min)
    
        
        if(startLocation < len(validationDataSequence_sales)):
            validationBatchInput = validationDataSequence_sales[startLocation: len(validationDataSequence_sales)]
            validationBatchTarget = validationDataSequence_sales_target[startLocation: len(validationDataSequence_sales)]
            
            validationLsBatch, validationForecastBatch = sess.run([loss, forecast], feed_dict={inputSequencePlaceholder: validationBatchInput,
                                                                                              targetPlaceholder: validationBatchTarget})
            
            totalValidationLoss+= validationLsBatch
        
        #validation completed.
        print('Validation completed after epoch:',e+1 ,'. Total validation loss:', totalValidationLoss)
        
    # Save the model
    print('----------- Saving Model')
    tf.saved_model.simple_save(sess, export_dir=TRAINED_MODEL_PATH, 
                    inputs={"inputSequencePlaceholder": inputSequencePlaceholder, "targetPlaceholder":targetPlaceholder}, 
                    outputs={"loss": loss, "forecast_originalScale": forecast_originalScale})
    print('saved model to:',TRAINED_MODEL_PATH )
    
print('----------- Finis')
        
            
    

    

starting training for epoch: 1
epoch: 1  iteration: 1
got a loss of: 0.75377196
the forecast of first 5 normalised are: [[0.99473137]
 [1.0000477 ]
 [0.9990285 ]
 [0.99471855]
 [0.9996665 ]]
while the actuals were normalised     : [[0.1199262 ]
 [0.07380074]
 [0.28413284]
 [0.        ]
 [0.06826568]]
the forecast of first 5 orignal scale are: [[539.1444 ]
 [542.0258 ]
 [541.47345]
 [539.13745]
 [541.8193 ]]
while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 1  iteration: 2
got a loss of: 6798.3623
the forecast of first 5 normalised are: [[-82.30445]
 [-82.31001]
 [-82.30894]
 [-82.30746]
 [-82.30726]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[-44609.01 ]
 [-44612.027]
 [-44611.445]
 [-44610.64 ]
 [-44610.535]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 1  iteration: 3
got a loss of: 70

while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 4  iteration: 2
got a loss of: 0.016394451
the forecast of first 5 normalised are: [[0.22375023]
 [0.10556412]
 [0.11208022]
 [0.18609416]
 [0.14489615]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[121.27263 ]
 [ 57.21575 ]
 [ 60.74748 ]
 [100.86304 ]
 [ 78.533714]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 4  iteration: 3
got a loss of: 0.01706857
the forecast of first 5 normalised are: [[0.10662365]
 [0.04147005]
 [0.04802883]
 [0.09224832]
 [0.04152346]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[57.790016]
 [22.476768]
 [26.031624]
 [49.99859 ]
 [22.505714]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 6

got a loss of: 0.014954215
the forecast of first 5 normalised are: [[0.19418216]
 [0.10093117]
 [0.105281  ]
 [0.16798627]
 [0.13488364]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[105.24673 ]
 [ 54.704693]
 [ 57.062298]
 [ 91.04856 ]
 [ 73.10693 ]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 7  iteration: 3
got a loss of: 0.019003548
the forecast of first 5 normalised are: [[0.16908443]
 [0.11495316]
 [0.11939871]
 [0.15717578]
 [0.11497641]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[91.64376]
 [62.30461]
 [64.7141 ]
 [85.18927]
 [62.31721]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 7  iteration: 4
got a loss of: 0.021131651
the forecast of first 5 normalised are: [

got a loss of: 0.016819535
the forecast of first 5 normalised are: [[0.1459111 ]
 [0.10339415]
 [0.10778189]
 [0.13764822]
 [0.10342586]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[79.08382 ]
 [56.03963 ]
 [58.41778 ]
 [74.60534 ]
 [56.056816]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 10  iteration: 4
got a loss of: 0.017748686
the forecast of first 5 normalised are: [[0.10717392]
 [0.10211849]
 [0.12748313]
 [0.1363275 ]
 [0.10227036]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[58.088264]
 [55.34822 ]
 [69.095856]
 [73.88951 ]
 [55.43054 ]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 10  iteration: 5
got a loss of: 0.0212436
the forecast of first 5 normalised are: [[0.14

got a loss of: 0.014242316
the forecast of first 5 normalised are: [[0.17558563]
 [0.12092686]
 [0.12471271]
 [0.16223764]
 [0.14390278]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[95.16741 ]
 [65.54236 ]
 [67.594284]
 [87.9328  ]
 [77.99531 ]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 13  iteration: 3
got a loss of: 0.017338658
the forecast of first 5 normalised are: [[0.15921271]
 [0.12343705]
 [0.12916279]
 [0.15454948]
 [0.12349701]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[86.29329 ]
 [66.90288 ]
 [70.00623 ]
 [83.765816]
 [66.93538 ]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 13  iteration: 4
got a loss of: 0.01750942
the forecast of first 5 normalised are: 

got a loss of: 0.014131594
the forecast of first 5 normalised are: [[0.18916512]
 [0.13507032]
 [0.13990879]
 [0.1733942 ]
 [0.15948033]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[102.5275  ]
 [ 73.208115]
 [ 75.83057 ]
 [ 93.97966 ]
 [ 86.43834 ]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 16  iteration: 3
got a loss of: 0.017760195
the forecast of first 5 normalised are: [[0.17364752]
 [0.13742054]
 [0.14854324]
 [0.17502165]
 [0.13756895]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[94.11696]
 [74.48193]
 [80.51044]
 [94.86173]
 [74.56237]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 16  iteration: 4
got a loss of: 0.017075641
the forecast of first 5 normalised are:

got a loss of: 0.0208044
the forecast of first 5 normalised are: [[0.20277965]
 [0.1667713 ]
 [0.18157995]
 [0.20850813]
 [0.16697788]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[109.90657 ]
 [ 90.39004 ]
 [ 98.41633 ]
 [113.011406]
 [ 90.502014]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 19  iteration: 4
got a loss of: 0.016891524
the forecast of first 5 normalised are: [[0.13610971]
 [0.12735462]
 [0.15033269]
 [0.14798212]
 [0.12815535]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[73.77146 ]
 [69.02621 ]
 [81.480316]
 [80.20631 ]
 [69.4602  ]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 19  iteration: 5
got a loss of: 0.021557102
the forecast of first 5 normalised are: [

epoch: 22  iteration: 4
got a loss of: 0.016412558
the forecast of first 5 normalised are: [[0.13246739]
 [0.12355399]
 [0.14572573]
 [0.14275408]
 [0.12434757]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[71.797325]
 [66.96626 ]
 [78.983345]
 [77.37271 ]
 [67.396385]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 22  iteration: 5
got a loss of: 0.01992784
the forecast of first 5 normalised are: [[0.13210046]
 [0.12217987]
 [0.12584364]
 [0.1293776 ]
 [0.1423086 ]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[71.59845 ]
 [66.22149 ]
 [68.20725 ]
 [70.12266 ]
 [77.131256]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 22  iteration: 6
got a loss of: 0.018015852
the forecast of first

got a loss of: 0.024551176
the forecast of first 5 normalised are: [[0.08286917]
 [0.06439388]
 [0.07964396]
 [0.07779455]
 [0.09881091]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[44.915092]
 [34.90148 ]
 [43.16703 ]
 [42.164646]
 [53.555515]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 25  iteration: 6
got a loss of: 0.022744665
the forecast of first 5 normalised are: [[0.07981324]
 [0.07096457]
 [0.05839348]
 [0.04484904]
 [0.08469737]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[43.258778]
 [38.4628  ]
 [31.649265]
 [24.308178]
 [45.90597 ]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 25  iteration: 7
got a loss of: 0.028394165
the forecast of first 5 normalised are:

got a loss of: 0.019541187
the forecast of first 5 normalised are: [[0.12756824]
 [0.10401714]
 [0.12434077]
 [0.12192726]
 [0.1435647 ]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[69.14199 ]
 [56.37729 ]
 [67.3927  ]
 [66.08458 ]
 [77.812065]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 28  iteration: 6
got a loss of: 0.020128684
the forecast of first 5 normalised are: [[0.09938943]
 [0.09125054]
 [0.07731032]
 [0.06369615]
 [0.10481095]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[53.869072]
 [49.45779 ]
 [41.902195]
 [34.52331 ]
 [56.807537]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 28  iteration: 7
got a loss of: 0.030129282
the forecast of first 5 normalised are:

got a loss of: 0.021096861
the forecast of first 5 normalised are: [[0.09010458]
 [0.07759917]
 [0.06315184]
 [0.04662764]
 [0.09468734]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[48.83668 ]
 [42.05875 ]
 [34.228294]
 [25.27218 ]
 [51.32054 ]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 31  iteration: 7
got a loss of: 0.027814934
the forecast of first 5 normalised are: [[0.10935795]
 [0.08906257]
 [0.12147868]
 [0.08421087]
 [0.06346536]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[59.27201 ]
 [48.271915]
 [65.841446]
 [45.642292]
 [34.398224]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 31  iteration: 7
got a loss of: 0.010155907
the forecast of first 5 normalised are:

got a loss of: 0.028947422
the forecast of first 5 normalised are: [[0.10143673]
 [0.08088672]
 [0.11357975]
 [0.07604158]
 [0.05488944]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[54.97871 ]
 [43.840603]
 [61.560226]
 [41.214535]
 [29.750076]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 34  iteration: 7
got a loss of: 0.010111099
the forecast of first 5 normalised are: [[0.17863429]
 [0.17167008]
 [0.1852076 ]
 [0.17507696]
 [0.18751657]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[ 96.819786]
 [ 93.04518 ]
 [100.38252 ]
 [ 94.891716]
 [101.63398 ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.24257994]
 [0.24475133]
 [0.2403127 ]

got a loss of: 0.011017306
the forecast of first 5 normalised are: [[0.1465553 ]
 [0.13282382]
 [0.15189135]
 [0.13953793]
 [0.15245569]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[79.432976]
 [71.99052 ]
 [82.32511 ]
 [75.629555]
 [82.63098 ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.21422136]
 [0.21684265]
 [0.21448898]
 [0.19142401]
 [0.21180224]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[116.10798 ]
 [117.52872 ]
 [116.25303 ]
 [103.751816]
 [114.796814]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 37 . Total validation loss: 0.05482510756701231
starting training for epoch: 38
epoch: 38  iteration: 1
got a loss

got a loss of: 0.010433956
the forecast of first 5 normalised are: [[0.15186656]
 [0.1430813 ]
 [0.1576941 ]
 [0.14593506]
 [0.15954626]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[82.311676]
 [77.55007 ]
 [85.4702  ]
 [79.0968  ]
 [86.47407 ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.21233726]
 [0.21536016]
 [0.2121427 ]
 [0.18583071]
 [0.21000504]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[115.08679 ]
 [116.72521 ]
 [114.981346]
 [100.720245]
 [113.82273 ]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 40 . Total validation loss: 0.05354719050228596
starting training for epoch: 41
epoch: 41  iteration: 1
got a loss

got a loss of: 0.02630455
the forecast of first 5 normalised are: [[0.11678183]
 [0.09467554]
 [0.132159  ]
 [0.08861208]
 [0.06729889]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[63.295753]
 [51.314144]
 [71.63017 ]
 [48.02775 ]
 [36.475998]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 43  iteration: 7
got a loss of: 0.010671534
the forecast of first 5 normalised are: [[0.20327806]
 [0.20456576]
 [0.21040964]
 [0.20066857]
 [0.21487355]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[110.17671 ]
 [110.87464 ]
 [114.04202 ]
 [108.76237 ]
 [116.461464]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.2600969 ]
 [0.26329494]
 [0.2579999 ]


epoch: 46  iteration: 7
got a loss of: 0.010729875
the forecast of first 5 normalised are: [[0.19948244]
 [0.19499838]
 [0.20595229]
 [0.19543111]
 [0.20864427]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[108.119484]
 [105.689125]
 [111.62614 ]
 [105.92366 ]
 [113.0852  ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.26545906]
 [0.26828384]
 [0.2643988 ]
 [0.23449183]
 [0.2615614 ]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[143.87881]
 [145.40985]
 [143.30415]
 [127.09457]
 [141.76628]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 46 . Total validation loss: 0.050261231139302254
starting training for epoch: 47
epoch: 47

first five predictions: [[0.24949372]
 [0.2522546 ]
 [0.24826467]
 [0.21697998]
 [0.24515426]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[135.2256 ]
 [136.722  ]
 [134.55945]
 [117.60315]
 [132.87361]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 49 . Total validation loss: 0.049935175105929375
starting training for epoch: 50
epoch: 50  iteration: 1
got a loss of: 0.022633351
the forecast of first 5 normalised are: [[0.26447213]
 [0.24406421]
 [0.26880944]
 [0.22300434]
 [0.21576607]]
while the actuals were normalised     : [[0.1199262 ]
 [0.07380074]
 [0.28413284]
 [0.        ]
 [0.06826568]]
the forecast of first 5 orignal scale are: [[143.34389 ]
 [132.2828  ]
 [145.69472 ]
 [120.868355]
 [116.94521 ]]
while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 50  iteration: 2
got a

while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 52 . Total validation loss: 0.04961380176246166
starting training for epoch: 53
epoch: 53  iteration: 1
got a loss of: 0.027852368
the forecast of first 5 normalised are: [[0.29144013]
 [0.27033627]
 [0.29796267]
 [0.24046695]
 [0.23662758]]
while the actuals were normalised     : [[0.1199262 ]
 [0.07380074]
 [0.28413284]
 [0.        ]
 [0.06826568]]
the forecast of first 5 orignal scale are: [[157.96056]
 [146.52226]
 [161.49576]
 [130.33308]
 [128.25215]]
while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 53  iteration: 2
got a loss of: 0.022455681
the forecast of first 5 normalised are: [[0.29000986]
 [0.21106553]
 [0.22026432]
 [0.25496447]
 [0.2519021 ]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[157.18535 ]
 [11

first five predictions: [[0.28674936]
 [0.28976524]
 [0.28461576]
 [0.24875355]
 [0.28181088]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[155.41815]
 [157.05276]
 [154.26173]
 [134.82442]
 [152.7415 ]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 55 . Total validation loss: 0.0513836033642292
starting training for epoch: 56
epoch: 56  iteration: 1
got a loss of: 0.031838957
the forecast of first 5 normalised are: [[0.3045987 ]
 [0.28387392]
 [0.3107649 ]
 [0.25632036]
 [0.25209713]]
while the actuals were normalised     : [[0.1199262 ]
 [0.07380074]
 [0.28413284]
 [0.        ]
 [0.06826568]]
the forecast of first 5 orignal scale are: [[165.09248]
 [153.85966]
 [168.43459]
 [138.92563]
 [136.63664]]
while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 56  iteration: 2
got a loss o

got a loss of: 0.033703756
the forecast of first 5 normalised are: [[0.3102926 ]
 [0.29044545]
 [0.3171481 ]
 [0.26163816]
 [0.25901234]]
while the actuals were normalised     : [[0.1199262 ]
 [0.07380074]
 [0.28413284]
 [0.        ]
 [0.06826568]]
the forecast of first 5 orignal scale are: [[168.17859]
 [157.42143]
 [171.89427]
 [141.80789]
 [140.38469]]
while the actuals were original scale     : [[ 65.]
 [ 40.]
 [154.]
 [  0.]
 [ 37.]]
epoch: 59  iteration: 2
got a loss of: 0.021887857
the forecast of first 5 normalised are: [[0.2832812 ]
 [0.21044981]
 [0.21824753]
 [0.25097942]
 [0.24725962]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[153.53842]
 [114.0638 ]
 [118.29016]
 [136.03085]
 [134.01471]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 59  iteration: 3
got a loss of: 0.014422494
the forecast of first 5 normalised are:

got a loss of: 0.019420166
the forecast of first 5 normalised are: [[0.27035773]
 [0.19749248]
 [0.20511019]
 [0.2375722 ]
 [0.23409057]]
while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[146.53389 ]
 [107.040924]
 [111.16972 ]
 [128.76413 ]
 [126.87709 ]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 62  iteration: 3
got a loss of: 0.01336457
the forecast of first 5 normalised are: [[0.1420387 ]
 [0.0983789 ]
 [0.12323999]
 [0.16302752]
 [0.09867954]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[76.98498 ]
 [53.32136 ]
 [66.796074]
 [88.360916]
 [53.484314]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 62  iteration: 4
got a loss of: 0.022478886
the forecast of first 5 normalised 

while the actuals were normalised     : [[0.19741697]
 [0.        ]
 [0.08118081]
 [0.11439114]
 [0.23800738]]
the forecast of first 5 orignal scale are: [[135.07143 ]
 [ 95.624794]
 [ 99.649445]
 [117.079926]
 [115.313965]]
while the actuals were original scale     : [[107.]
 [  0.]
 [ 44.]
 [ 62.]
 [129.]]
epoch: 65  iteration: 3
got a loss of: 0.013357598
the forecast of first 5 normalised are: [[0.11584449]
 [0.07272363]
 [0.09728706]
 [0.13722944]
 [0.07301748]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[62.787712]
 [39.416206]
 [52.729588]
 [74.37836 ]
 [39.575474]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 65  iteration: 4
got a loss of: 0.02543585
the forecast of first 5 normalised are: [[ 0.00117493]
 [-0.0111289 ]
 [ 0.01320744]
 [ 0.01115167]
 [-0.00995076]]
while the actuals were normalised     : [[0.11808118]
 [0

got a loss of: 0.014790747
the forecast of first 5 normalised are: [[0.08756828]
 [0.04406524]
 [0.06956553]
 [0.1106174 ]
 [0.04436636]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[47.46201 ]
 [23.883358]
 [37.70452 ]
 [59.95463 ]
 [24.046566]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 68  iteration: 4
got a loss of: 0.027144367
the forecast of first 5 normalised are: [[-0.0068928 ]
 [-0.01981676]
 [ 0.00469637]
 [ 0.00208068]
 [-0.0185622 ]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[ -3.7358978]
 [-10.740682 ]
 [  2.545432 ]
 [  1.127728 ]
 [-10.060711 ]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 68  iteration: 5
got a loss of: 0.0345664
the forecast of first 5 normali

got a loss of: 0.017032124
the forecast of first 5 normalised are: [[0.06422591]
 [0.01791036]
 [0.04731536]
 [0.0919013 ]
 [0.01825869]]
while the actuals were normalised     : [[0.16236162]
 [0.        ]
 [0.1199262 ]
 [0.22693727]
 [0.        ]]
the forecast of first 5 orignal scale are: [[34.810444]
 [ 9.707416]
 [25.644924]
 [49.810505]
 [ 9.896211]]
while the actuals were original scale     : [[ 88.]
 [  0.]
 [ 65.]
 [123.]
 [  0.]]
epoch: 71  iteration: 4
got a loss of: 0.024911232
the forecast of first 5 normalised are: [[ 0.00207174]
 [-0.01323426]
 [ 0.01313698]
 [ 0.00827885]
 [-0.01168811]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[ 1.1228821]
 [-7.172968 ]
 [ 7.120245 ]
 [ 4.487135 ]
 [-6.334957 ]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 71  iteration: 5
got a loss of: 0.025271222
the forecast of first 5 normalised

got a loss of: 0.016076254
the forecast of first 5 normalised are: [[0.05150914]
 [0.02711093]
 [0.0619266 ]
 [0.0486815 ]
 [0.02982438]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[27.917955]
 [14.694126]
 [33.56422 ]
 [26.385372]
 [16.164812]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 74  iteration: 5
got a loss of: 0.015660545
the forecast of first 5 normalised are: [[0.11956358]
 [0.03337562]
 [0.14502525]
 [0.10099471]
 [0.18359971]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[64.80346 ]
 [18.089586]
 [78.60368 ]
 [54.739132]
 [99.51104 ]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 74  iteration: 6
got a loss of: 0.016844328
the forecast of first 5 normalised are: [[0.

got a loss of: 0.012225053
the forecast of first 5 normalised are: [[0.09047043]
 [0.05032444]
 [0.10085988]
 [0.07465005]
 [0.05529356]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[49.034973]
 [27.275846]
 [54.666054]
 [40.460327]
 [29.969109]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 77  iteration: 5
got a loss of: 0.014161427
the forecast of first 5 normalised are: [[0.12322783]
 [0.02163625]
 [0.16174996]
 [0.0970552 ]
 [0.20700991]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[ 66.78949 ]
 [ 11.726847]
 [ 87.66848 ]
 [ 52.603916]
 [112.19937 ]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 77  iteration: 6
got a loss of: 0.01423553
the forecast of first 5 normalised are: 

got a loss of: 0.012132516
the forecast of first 5 normalised are: [[0.13710666]
 [0.07657444]
 [0.14467192]
 [0.10466051]
 [0.08402383]]
while the actuals were normalised     : [[0.11808118]
 [0.        ]
 [0.06457565]
 [0.00184502]
 [0.02398524]]
the forecast of first 5 orignal scale are: [[74.311806]
 [41.50335 ]
 [78.41218 ]
 [56.725998]
 [45.540916]]
while the actuals were original scale     : [[64.]
 [ 0.]
 [35.]
 [ 1.]
 [13.]]
epoch: 80  iteration: 5
got a loss of: 0.016039714
the forecast of first 5 normalised are: [[ 0.07793045]
 [-0.02256286]
 [ 0.1280495 ]
 [ 0.04890621]
 [ 0.18458235]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[ 42.238304]
 [-12.229071]
 [ 69.402824]
 [ 26.507164]
 [100.04363 ]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 80  iteration: 6
got a loss of: 0.014185441
the forecast of first 5 normalised

got a loss of: 0.015325191
the forecast of first 5 normalised are: [[0.10420775]
 [0.01941144]
 [0.14052463]
 [0.08307016]
 [0.18733919]]
while the actuals were normalised     : [[0.07749077]
 [0.        ]
 [0.33210332]
 [0.        ]
 [0.22324723]]
the forecast of first 5 orignal scale are: [[ 56.480602]
 [ 10.521003]
 [ 76.164345]
 [ 45.024025]
 [101.53784 ]]
while the actuals were original scale     : [[ 42.]
 [  0.]
 [180.]
 [  0.]
 [121.]]
epoch: 83  iteration: 6
got a loss of: 0.014243779
the forecast of first 5 normalised are: [[0.20644796]
 [0.13526917]
 [0.14473224]
 [0.10382271]
 [0.18466711]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[111.89479 ]
 [ 73.31589 ]
 [ 78.44487 ]
 [ 56.271908]
 [100.08958 ]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 83  iteration: 7
got a loss of: 0.020526454
the forecast of first 5 norma

got a loss of: 0.013848692
the forecast of first 5 normalised are: [[0.20396316]
 [0.12582088]
 [0.13965082]
 [0.09550834]
 [0.17986548]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[110.548035]
 [ 68.194916]
 [ 75.69074 ]
 [ 51.76552 ]
 [ 97.48709 ]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 86  iteration: 7
got a loss of: 0.019419337
the forecast of first 5 normalised are: [[0.19362974]
 [0.1531856 ]
 [0.2336185 ]
 [0.13214707]
 [0.11102664]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[104.94732 ]
 [ 83.026596]
 [126.62122 ]
 [ 71.62371 ]
 [ 60.17644 ]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 86  iteration: 7
got a loss of: 0.009105901
the forecast of first 5 norma

epoch: 89  iteration: 7
got a loss of: 0.018423324
the forecast of first 5 normalised are: [[0.18447006]
 [0.13857436]
 [0.23333871]
 [0.11395323]
 [0.0932951 ]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[ 99.98277 ]
 [ 75.10731 ]
 [126.46958 ]
 [ 61.762653]
 [ 50.56594 ]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 89  iteration: 7
got a loss of: 0.008806742
the forecast of first 5 normalised are: [[0.19921458]
 [0.25093663]
 [0.20450783]
 [0.19664693]
 [0.22480857]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[107.974304]
 [136.00764 ]
 [110.84325 ]
 [106.582634]
 [121.846245]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.15700877

got a loss of: 0.01334485
the forecast of first 5 normalised are: [[0.17938542]
 [0.07314205]
 [0.10365319]
 [0.04637027]
 [0.14462078]]
while the actuals were normalised     : [[0.61808118]
 [0.0498155 ]
 [0.20479705]
 [0.14206642]
 [0.17343173]]
the forecast of first 5 orignal scale are: [[97.2269  ]
 [39.64299 ]
 [56.18003 ]
 [25.132685]
 [78.38446 ]]
while the actuals were original scale     : [[335.]
 [ 27.]
 [111.]
 [ 77.]
 [ 94.]]
epoch: 92  iteration: 7
got a loss of: 0.0193766
the forecast of first 5 normalised are: [[0.14719105]
 [0.0999105 ]
 [0.19782996]
 [0.07180452]
 [0.05218744]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[ 79.77755 ]
 [ 54.15149 ]
 [107.22384 ]
 [ 38.918053]
 [ 28.285595]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 92  iteration: 7
got a loss of: 0.009575805
the forecast of first 5 normalised ar

got a loss of: 0.018322652
the forecast of first 5 normalised are: [[0.15520382]
 [0.1056819 ]
 [0.21275187]
 [0.07724547]
 [0.05857694]]
while the actuals were normalised     : [[0.11439114]
 [0.09225092]
 [0.25276753]
 [0.        ]
 [0.        ]]
the forecast of first 5 orignal scale are: [[ 84.12047 ]
 [ 57.279587]
 [115.31151 ]
 [ 41.867046]
 [ 31.748703]]
while the actuals were original scale     : [[ 62.]
 [ 50.]
 [137.]
 [  0.]
 [  0.]]
epoch: 95  iteration: 7
got a loss of: 0.009380517
the forecast of first 5 normalised are: [[0.20285988]
 [0.27565074]
 [0.20688593]
 [0.2011776 ]
 [0.2319572 ]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[109.95006]
 [149.4027 ]
 [112.13218]
 [109.03825]
 [125.7208 ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.15848339]
 [0.17118835]
 [0.16165447]

got a loss of: 0.008812735
the forecast of first 5 normalised are: [[0.1994586 ]
 [0.26744223]
 [0.20359242]
 [0.19883978]
 [0.22777426]]
while the actuals were normalised     : [[0.19741697]
 [0.17343173]
 [0.099631  ]
 [0.20479705]
 [0.17896679]]
the forecast of first 5 orignal scale are: [[108.10656 ]
 [144.95369 ]
 [110.34709 ]
 [107.771164]
 [123.45365 ]]
while the actuals were original scale     : [[107.]
 [ 94.]
 [ 54.]
 [111.]
 [ 97.]]
starting validation
first five predictions: [[0.15620553]
 [0.16873801]
 [0.1590035 ]
 [0.11163092]
 [0.17942262]]
first five actuals    : [[0.15682657]
 [0.14391144]
 [0.13099631]
 [0.20848708]
 [0.25461255]]
the forecast of first 5 orignal scale are: [[84.6634  ]
 [91.456   ]
 [86.17989 ]
 [60.503956]
 [97.247055]]
while the actuals were original scale     : [[ 85.]
 [ 78.]
 [ 71.]
 [113.]
 [138.]]
Validation completed after epoch: 98 . Total validation loss: 0.04394811391830444
starting training for epoch: 99
epoch: 99  iteration: 1
got a loss

INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: /home/ubuntu/anurag/rnn/savedModel/saved_model.pb
saved model to: /home/ubuntu/anurag/rnn/savedModel
----------- Finis
